src/gen/async_generate_data_gemini.py

```diff
@@ -0,0 +1,99 @@
+RATE_LIMIT_PER_MINUTE = 60
+REQUEST_INTERVAL = 60 / RATE_LIMIT_PER_MINUTE  # Seconds between requests
+
+
+# prompt template can be loaded from a `.toml` file.
+def craft_prompt(instruction, response):
+    prompt = """
+Generate a series of (instruction, response) pairs that are similar in context and structure to the example provided below. Each pair should consist of a concise instruction followed by an appropriate, detailed response. The instruction should pose a clear task or question, while the response should provide a comprehensive answer or solution that could be understood by someone with a basic understanding of the subject.
+
+Example pair:
+
+Instruction: {instruction}
+Response: {response}
+
+Your task is to generate more pairs that maintain this level of clarity and detail. The topic is Coding. Ensure that the responses are informative and accurate, suitable for an educational context.
+
+Store the generated pairs in JSON format, with each pair as an object within an array. Each object should have two key-value pairs: "instruction" and "response". For instance:
+
+[{{"instruction": "text", "response": "text"}}, {{"instruction": "text", "response": "text"}}, ...]
+
+Remember to maintain consistency in the format and ensure the generated pairs are diverse and cover a broad range of subjects. You must return the response
+in the asked format and you must not add any additional text in your response.
+"""
+    return prompt.format(instruction=instruction, response=response)
+
+
+async def generate_text(prompt):
+    url = (
+        url
+    ) = f"https://generativelanguage.googleapis.com/v1beta/models/{MODEL_NAME}:generateContent?key={GOOGLE_API_KEY}"
+    data = {"contents": [{"parts": [{"text": prompt}]}]}
+
+    async with aiohttp.ClientSession() as session:
+        async with session.post(url, json=data) as response:
+            if response.status == 200:
+                result = await response.json()
+                return result
+            else:
+                print(f"Error: {response.status}")
+                return None
+
+
+async def main():
+    # craft prompts
+    total_original_samples = len(EXISTING_DATASET)
+    random_indices = np.random.randint(0, total_original_samples, size=(NUM_DATASET_SAMPLES))
+    prompts = []
+    print("Preparing candidate prompts.")
+    for random_index in random_indices:
+        sample = EXISTING_DATASET[int(random_index)]["messages"]
+        instruction = sample[0]["content"]
+        response = sample[1]["content"]
+        prompt_for_sample = craft_prompt(instruction, response)
+        prompts.append(prompt_for_sample)
+
+    # create tasks
+    tasks = []
+    print("Creating tasks for asyncio.")
+    for prompt in prompts:
+        tasks.append(generate_text(prompt))
+        # rate-limiting
+        await asyncio.sleep(REQUEST_INTERVAL)
+
+    # fire execution
+    results = await asyncio.gather(*tasks)
```

Comment:
maybe I am wrong, but AFAIK, when `asyncio.gather` is called ans awaited the tasks will be scheduled to be executed ([ref](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather)). That means `await asyncio.sleep(REQUEST_INTERVAL)` does not work as expected.

Reply:
```python
import asyncio

async def factorial(name, number):
    f = 1
    for i in range(2, number + 1):
        print(f"Task {name}: Compute factorial({number}), currently i={i}...")
        await asyncio.sleep(1)
        f *= i
    print(f"Task {name}: factorial({number}) = {f}")
    return f

async def main():
    # Schedule three calls *concurrently*:
    L = await asyncio.gather(
        factorial("A", 2),
        factorial("B", 3),
        factorial("C", 4),
    )
    print(L)

asyncio.run(main())
```

The official example above has `await asyncio.sleep()` within the task to be executed. So, I think it's okay. If not, please be more specific.

Reply:
To be clearer,

```python
tasks = []
for prompt in prompts:
    tasks.append(generate_text(prompt))
    await asyncio.sleep(1)
```

the above code snippet won't execute `generate_text()` but sleeping for `len(prompts)` since `generate_text()` hasn't been `awaited`. 

I think the all tasks appended into the tasks will be executed when `await asyncio.gather(*tasks)` is called since that is where `await` happens. So, if your purpose was "execute a async jobs" -> "sleep" -> "execute a async jobs" -> "sleep" -> ..., that is not going to happen?

correct me if I am wrong

Reply:
I see. If that is the case `await generate_text()` should cut it no?

Reply:
I think of the two possible implementation choices

- keep `asyncio.gather`. and make `asyncio.sleep(1)` happen inside `generate_text()`
- remove `asyncio.gather`. fire `await generate_text()` in the loop, but calculate the number of requests every minute.